cfd_protocol/src/interval.rs

```diff
@@ -0,0 +1,124 @@
+use anyhow::{bail, Result};
+use bit_vec::BitVec;
+use std::fmt::Display;
+use std::ops::RangeInclusive;
+
+mod digit_decomposition;
+
+const BASE: usize = 2;
+
+/// Maximum number of binary digits for BTC price in whole USD.
+const MAX_DIGITS: usize = 20;
+
+const MAX_PRICE_DEC: u64 = (BASE as u64).pow(MAX_DIGITS as u32);
+
+#[derive(Debug)]
+pub struct Interval(RangeInclusive<u64>);
+
+impl Interval {
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > MAX_PRICE_DEC || end > MAX_PRICE_DEC {
+            bail!("price over maximum")
+        }
+
+        if start > end {
+            bail!("invalid interval: start > end")
+        }
+
+        Ok(Self(start..=end))
+    }
+
+    pub fn as_digits(&self) -> Vec<Digits> {
```

Comment:
This allocates :upside_down_face: 

```suggestion
    pub fn to_digits(&self) -> Vec<Digits> {
```

Reply:
I promise I will not mess this up one day.